### PR TITLE
WebSocket based debugger interface

### PR DESCRIPTION
**Pull request by unknownbrackets**
If you want to try it:

1. Run this build on your device (Android / iOS / desktop / whatever.)
2. Enable Settings -> Tools -> Developer tools -> Allow remote debugger.
3. Go here: http://ppsspp-debugger.unknownbrackets.org/

Alternatively (only logs, no debugger):
Copy this into an HTML file (desktop, whatever): https://gist.github.com/unknownbrackets/385c01ca12a5ac22071f688e84c39ba3 (possibly PPSSPP could embed its own default HTML...)
Open the HTML file in a web browser and play a game: watch the logs roll in.

This is just a thrown-together example and of course command line, lua, whatever could be used to speak to this WebSocket ([a list of WebSocket libraries](https://github.com/facundofarias/awesome-websockets).)  I'm planning a simple JSON API.

The debugger would also work fine in a Chrome tab on the device (assuming a mobile friendly debugger UI...) as long as PPSSPP is not quit (which might become the case on iOS...)

 - [x] Update json #10960
 - [x] Fix debugger threading #10987
 - [x] IPv6 support #11004
 - [x] Include armips everywhere #11093
 - [x] Breakpoints on ARM #11148
 - [x] Remove Qt debugger #11130
 - [x] WebSocket / separated webserver #11152
 - [x] Rebase this (possibly merge some small commits together after rebase)

-[Unknown]

### CODE REVIEW

File: UI/RemoteISOScreen.cpp [hunk 1]
```diff
@@ -161,6 +162,10 @@ static void ExecuteServer() {
 		http->RegisterHandler(pair.first.c_str(), handler);
 	}
 
+	// TODO: This isn't an ideal place to put this... should separate.
+	// But kinda nice to only have it listen on one port?
```
Comment (hrydgard) [inline UI/RemoteISOScreen.cpp]: Yeah, I think this is fine really.

**Comment by hrydgard** [comment]
Really cool stuff!

Regarding JSON serializer, we have a simple writer in ext/native/json_writer and a (different) reader in ext/native/ext/vjson. Might be worth using a proper one if we're gonna write a lot of different structs to json though, but I don't know much about which the best one today might be.

Same thing with React etc, not a frontend guy :)

File: UI/RemoteISOScreen.cpp (hunk 1 above)
Comment (unknownbrackets) [reply UI/RemoteISOScreen.cpp]: Cool.  Thinking of moving it to Core/WebServer.cpp or something, and sending flags to it to restart with remote iso / debugger enabled.

-[Unknown]

**Comment by unknownbrackets** [comment]
Okay, just using that json_writer then (totally forgot it existed.)  I think it will work fine.

Not sure if anyone else has opinions on frontend or where to source them - ideally this would make hacking, translation, cheat, whatever tools more accessible.

-[Unknown]

**Comment by unknownbrackets** [comment]
Went ahead and switched from vjson to gason (newer from same author.)  Otherwise, vjson would parse 32-bit unsigned integer parameters as either truncated ints or floats.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, gason seems fine. Good stuff.

**Comment by weituotian** [comment]
Really cool stuff!

if it is allow, i want `Memory::base`;╰(￣▽￣)╮

**Comment by unknownbrackets** [comment]
I'm planning to have memory APIs: read, write, disasm, assemble, and memory breakpoint management.

Are you asking for an API to return the C++ address in Memory::base?  Guess I can add that (although it'll only be useful on some platforms, I think.)

-[Unknown]

**Comment by unknownbrackets** [comment]
Just a note: at this point the branch supports other platforms (Linux, Android, etc.) but there's no breakpoint support on ARM.  That means you can only step into and step HLE on ARM devices at the moment.

It also exposes Memory.base for all your cheat searching needs.

-[Unknown]

**Comment by unknownbrackets** [comment]
Okay, I've published a somewhat functional debugger UI here:

http://ppsspp-debugger.unknownbrackets.org/

As noted above, x86 only for any breakpointing / step over / step out / etc.  You'll need to compile the latest branch here.

-[Unknown]

**Comment by unknownbrackets** [comment]
Now arm64 has breakpoints.  Didn't test memory breakpoints yet, but hopefully they work....

-[Unknown]

**Comment by unknownbrackets** [comment]
New debugger (the multiplatform web one) now supports thread context debugging.  Put simply, you can now view and change thread regs, or select a thread and click "step into" to (mostly) step into that thread.

It's still based on address breakpoints (no thread condition) so it could break to a different thread.  Still a good improvement.

-[Unknown]

**Comment by unknownbrackets** [comment]
@hrydgard just a request in case you start merging any of the 1.7.x pulls next week: would be great not to merge them too quickly so the buildbot builds each merge.

Also not this pull yet, I want to rebase it (hence WIP.)

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, I know what we use WIP in titles for :)

I'll probably start merging towards the end of next week, hopefully 1.6.3 will be the last in the series.

**Comment by unknownbrackets** [comment]
Heh, sorry.  I think it's good now - combined a few typo fixes into earlier commits.

At this point it's mostly usable as a debugger, although there's no memory view/search yet and nothing at all for GE debugger yet.

-[Unknown]

**Comment by hrydgard** [comment]
Cool. I'll try to take a look soon - it's kind of a big one :)

**Comment by unknownbrackets** [comment]
Heh, it's a couple few lines.  Happy to chop it up more if it helps, would be mostly linear at this point.

-[Unknown]

**Comment by hrydgard** [comment]
Might be neat to pull the http: commits into a separate pull request which we'll merge first, but I'm not gonna demand anything :)

**Comment by unknownbrackets** [comment]
Sure, separated out - #11152.

-[Unknown]

**Comment by hrydgard** [comment]
please rebase on master so we get the buildfix that was just committed in any future bisect :)

**Comment by unknownbrackets** [comment]
Yep, rebased.  I haven't tested these changes with GCC 8 either, though.

-[Unknown]

**Comment by hrydgard** [comment]
Well, I guess there will be enough platforms that we can bisect on either way :) 

I'll give this a quick read through a little later and then probably merge.

**Comment by hrydgard** [comment]
@unknownbrackets I'm fine with this. Again, it says "Not ready for merge" in the pull request, so should I or not? :)   If you're ready you can go ahead.

**Review by hrydgard — APPROVED** [review]

**Comment by unknownbrackets** [comment]
Oops, thanks for checking - forgot to remove it from there.

-[Unknown]

**Comment by leopard20** [comment]
@unknownbrackets Finally! Let's check this out! :)

**Comment by leopard20** [comment]
@unknownbrackets So...how do I make this work on Android again? I enabled Allow Remote Debugging but when I go to the mentioned site it says it can't connect to PPSSPP.

**Comment by unknownbrackets** [comment]
Hm.  That's all that should be needed.  Specifically:

1. Make sure both devices are connected to the same wifi and use NAT (e.g. you have local IPs like 10.x.x.x or 192.168.x.x or 172.16.x.x.)
2. Make sure Allow Remote Debugging is checked on the device you want to debug (e.g. Android.)
3. While PPSSPP is running (doesn't need to be in a game yet), open http://ppsspp-debugger.unknownbrackets.org/.

IMPORTANT: https://ppsspp-debugger.unknownbrackets.org/ will not work.  It unfortunately CANNOT be SSL because the emulator itself doesn't (and can't really) have a valid SSL certificate.

If this doesn't work, you can try looking at the IP and port your device is using (it can be seen under Tools -> Remote Disc Streaming -> Settings) and type it manually.  If you go to http://192.x.x.x:zzzz/ in your browser it should also show a text file with at least one line "/debugger".

If you can't access the ip:port from your PC browser, then it could mean something is blocking the connection (either not the same wifi, or your device has a firewall on it - most Androids don't.)  If you can, but the debugger doesn't work, it probably means a bug...

-[Unknown]

**Comment by unknownbrackets** [comment]
Just wondering - did that help?

Also, I've outlined the missing features for the web debugger UI here:
https://github.com/unknownbrackets/ppsspp-debugger/projects

-[Unknown]

**Comment by hrydgard** [comment]
Confirming that it works fine on Android for me if I have the device and the debugger on the same network. If it matters, whatismyip shows only a IPv4 address for me where I am now.

**Comment by unknownbrackets** [comment]
Ah, yes, it might be a little restrictive for ipv6 currently.  It currently assumes the devices will be on the same /64 or "ipv6 subnet", which is commonly the case.

Your internet provider may give you an entire /56, but usually your router will assign IPs within a single /64 ipv6 subnet even so.  If the router ends up assigning IPs more liberally and your mobile device and desktop are in entirely different subnets, then the automatic IP detection won't work.

-[Unknown]

**Comment by leopard20** [comment]
Derp, forgot about my Android device firewall! Now it works! Thanks!

**Pull request merged**